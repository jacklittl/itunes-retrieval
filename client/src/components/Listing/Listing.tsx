import React, { useState, useEffect } from "react";
import { Result } from '../../declarations/types';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';

import ListItem from '../ListItem/ListItem';

type Props = {
    results: Result[];
}

const Listing = ({ results }: Props): JSX.Element => {
    const [hasMoreResults, setHasMoreResults] = useState<boolean>(false);
    const [resultCount, setResultCount] = useState<number>(10);

    useEffect(() => setHasMoreResults( results.length > 10 ), [results]);

    const activeResults = results.slice( 0, resultCount );

    const loadMoreResults = () => {
        if ( resultCount >= results.length ) setHasMoreResults(false);
        else setResultCount(resultCount + 10);
    }

    window.onscroll = () => {
        var d = document.documentElement,
            offset = d.scrollTop + window.innerHeight;

        if (offset >= document.body.scrollHeight) {
            hasMoreResults && loadMoreResults();
        }
    }
    
    return (

        <>

            <Typography gutterBottom variant="h5" color="primary" component="h2">Search Results:</Typography>
            
            <Grid container spacing={2}>
                
                {activeResults.map( result => (

                    <Grid item xs={3} key={result.trackId}>

                        <ListItem result={result} />

                    </Grid>

                ))}

            </Grid>

        </>

    );
};

export default ( Listing )