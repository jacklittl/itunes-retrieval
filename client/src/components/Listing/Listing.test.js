import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from "enzyme";
import Adapter from 'enzyme-adapter-react-16';

import Listing from './Listing';

configure({adapter: new Adapter()});

const resultsPayload = [{
  artistId: 6906197,
  artistName: "Foo Fighters",
  artistViewUrl: "https://music.apple.com/us/artist/foo-fighters/6906197?uo=4",
  artworkUrl30: "https://is2-ssl.mzstatic.com/image/thumb/Music/v4/e4/23/35/e4233532-42dc-7ece-d18b-b9b2733cb4e6/source/30x30bb.jpg",
  artworkUrl60: "https://is2-ssl.mzstatic.com/image/thumb/Music/v4/e4/23/35/e4233532-42dc-7ece-d18b-b9b2733cb4e6/source/60x60bb.jpg",
  artworkUrl100: "https://is2-ssl.mzstatic.com/image/thumb/Music/v4/e4/23/35/e4233532-42dc-7ece-d18b-b9b2733cb4e6/source/100x100bb.jpg",
  collectionCensoredName: "In Your Honor",
  collectionExplicitness: "notExplicit",
  collectionId: 300855107,
  collectionName: "In Your Honor",
  collectionPrice: 16.99,
  collectionViewUrl: "https://music.apple.com/us/album/best-of-you/300855107?i=300855110&uo=4",
  country: "USA",
  currency: "USD",
  discCount: 2,
  discNumber: 1,
  isStreamable: true,
  kind: "song",
  previewUrl: "https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview71/v4/a1/34/a5/a134a568-a74a-042e-2c7e-1192efbac8aa/mzaf_8504368996424262545.plus.aac.p.m4a",
  primaryGenreName: "Alternative",
  releaseDate: "2005-05-30T07:00:00Z",
  trackCensoredName: "Best of You",
  trackCount: 10,
  trackExplicitness: "notExplicit",
  trackId: 300855110,
  trackName: "Best of You",
  trackNumber: 3,
  trackPrice: 1.29,
  trackTimeMillis: 255261,
  trackViewUrl: "https://music.apple.com/us/album/best-of-you/300855107?i=300855110&uo=4",
  wrapperType: "track"
}, {
  artistId: 6906197,
  artistName: "Foo Fighters",
  artistViewUrl: "https://music.apple.com/us/artist/foo-fighters/6906197?uo=4",
  artworkUrl30: "https://is5-ssl.mzstatic.com/image/thumb/Music3/v4/41/b0/26/41b02667-a47c-e969-71f6-918825311607/source/30x30bb.jpg",
  artworkUrl60: "https://is5-ssl.mzstatic.com/image/thumb/Music3/v4/41/b0/26/41b02667-a47c-e969-71f6-918825311607/source/60x60bb.jpg",
  artworkUrl100: "https://is5-ssl.mzstatic.com/image/thumb/Music3/v4/41/b0/26/41b02667-a47c-e969-71f6-918825311607/source/100x100bb.jpg",
  collectionCensoredName: "Greatest Hits",
  collectionExplicitness: "explicit",
  collectionId: 334811953,
  collectionName: "Greatest Hits",
  collectionPrice: 11.99,
  collectionViewUrl: "https://music.apple.com/us/album/everlong-acoustic-version/334811953?i=334812065&uo=4",
  country: "USA",
  currency: "USD",
  discCount: 1,
  discNumber: 1,
  isStreamable: true,
  kind: "song",
  previewUrl: "https://audio-ssl.itunes.apple.com/itunes-assets/Music/e5/ec/db/mzm.okcwwdvf.aac.p.m4a",
  primaryGenreName: "Alternative",
  releaseDate: "1997-05-20T07:00:00Z",
  trackCensoredName: "Everlong (Acoustic Version)",
  trackCount: 16,
  trackExplicitness: "notExplicit",
  trackId: 334812065,
  trackName: "Everlong",
  trackNumber: 16,
  trackPrice: 1.29,
  trackTimeMillis: 250852,
  trackViewUrl: "https://music.apple.com/us/album/everlong-acoustic-version/334811953?i=334812065&uo=4",
  wrapperType: "track"
}]

describe('Listing Render', () => {
  test('Listing Renders', () => {
    const component = renderer.create(
        <Listing results={resultsPayload} />
    );
    
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});