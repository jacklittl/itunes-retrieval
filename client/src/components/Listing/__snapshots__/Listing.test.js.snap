// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listing Render Listing Renders 1`] = `
Array [
  <h2
    className="MuiTypography-root MuiTypography-h5 MuiTypography-colorPrimary MuiTypography-gutterBottom"
  >
    Search Results:
  </h2>,
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          className="MuiButtonBase-root MuiCardActionArea-root"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            href="https://music.apple.com/us/album/best-of-you/300855107?i=300855110&uo=4"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="Best of You"
              className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
              src="https://is2-ssl.mzstatic.com/image/thumb/Music/v4/e4/23/35/e4233532-42dc-7ece-d18b-b9b2733cb4e6/source/100x100bb.jpg"
              title="Best of You"
            />
            <div
              className="MuiCardContent-root"
            >
              <h2
                className="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom"
              >
                Best of You
              </h2>
              <p
                className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              >
                Foo Fighters
              </p>
              <p
                className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
              >
                In Your Honor
              </p>
            </div>
          </a>
          <span
            className="MuiCardActionArea-focusHighlight"
          />
        </button>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          className="MuiButtonBase-root MuiCardActionArea-root"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            href="https://music.apple.com/us/album/everlong-acoustic-version/334811953?i=334812065&uo=4"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="Everlong"
              className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
              src="https://is5-ssl.mzstatic.com/image/thumb/Music3/v4/41/b0/26/41b02667-a47c-e969-71f6-918825311607/source/100x100bb.jpg"
              title="Everlong"
            />
            <div
              className="MuiCardContent-root"
            >
              <h2
                className="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom"
              >
                Everlong
              </h2>
              <p
                className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              >
                Foo Fighters
              </p>
              <p
                className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
              >
                Greatest Hits
              </p>
            </div>
          </a>
          <span
            className="MuiCardActionArea-focusHighlight"
          />
        </button>
      </div>
    </div>
  </div>,
]
`;

exports[`Listing Render snapshot renders 1`] = `
Array [
  <h2
    className="MuiTypography-root MuiTypography-h5 MuiTypography-colorPrimary MuiTypography-gutterBottom"
  >
    Search Results:
  </h2>,
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          className="MuiButtonBase-root MuiCardActionArea-root"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            href="https://music.apple.com/us/album/best-of-you/300855107?i=300855110&uo=4"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="Best of You"
              className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
              src="https://is2-ssl.mzstatic.com/image/thumb/Music/v4/e4/23/35/e4233532-42dc-7ece-d18b-b9b2733cb4e6/source/100x100bb.jpg"
              title="Best of You"
            />
            <div
              className="MuiCardContent-root"
            >
              <h2
                className="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom"
              >
                Best of You
              </h2>
              <p
                className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              >
                Foo Fighters
              </p>
              <p
                className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
              >
                In Your Honor
              </p>
            </div>
          </a>
          <span
            className="MuiCardActionArea-focusHighlight"
          />
        </button>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          className="MuiButtonBase-root MuiCardActionArea-root"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
            href="https://music.apple.com/us/album/everlong-acoustic-version/334811953?i=334812065&uo=4"
            onBlur={[Function]}
            onFocus={[Function]}
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="Everlong"
              className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
              src="https://is5-ssl.mzstatic.com/image/thumb/Music3/v4/41/b0/26/41b02667-a47c-e969-71f6-918825311607/source/100x100bb.jpg"
              title="Everlong"
            />
            <div
              className="MuiCardContent-root"
            >
              <h2
                className="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom"
              >
                Everlong
              </h2>
              <p
                className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              >
                Foo Fighters
              </p>
              <p
                className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
              >
                Greatest Hits
              </p>
            </div>
          </a>
          <span
            className="MuiCardActionArea-focusHighlight"
          />
        </button>
      </div>
    </div>
  </div>,
]
`;
